M2/Macaulay2/packages/MatrixSchubert/permutationMethods.m2

```diff
@@ -0,0 +1,569 @@
+    if (isIdentity w) then return 1;
+    r := lastDescent(w) - 1;
+    s := -1;
+    scan(reverse(r+1..n-1), i -> if w_i < w_r then (s = i; break));
+    v := swap(w,r,s);
+    previnds := select(0..r-1, q -> permLength(swap(v,q,r)) == permLength(v)+1);
+    us := apply(previnds, i -> swap(v, i, r));
+    if not isDouble then
+        sum(toList(apply(us, u -> schubertPolyHelper(u, Q, Double=>isDouble)))) + Q_r * schubertPolyHelper(v, Q, Double=>isDouble)
+    else 
+        sum(toList(apply(us, u -> schubertPolyHelper(u, Q, Double=>isDouble)))) + (Q_r - Q_(n-1+v_r)) * schubertPolyHelper(v, Q, Double=>isDouble)
+)
+
+schubertPolynomial = method(Options=>{Algorithm=>"DividedDifference"})
+schubertPolynomial(List) := opts-> (w) -> (
+    
+    n := #w;
+    x := local x;
+    Q := QQ[x_1..x_n];
+    if opts.Algorithm == "Transition" then 
+        schubertPolyHelper(w, Q, Double=>false)
+    else if opts.Algorithm == "DividedDifference" then
+        polyByDividedDifference(w,Q)
+    else error("Invalid option for Algorithm.")
+    
+)
+
+doubleSchubertPolynomial = method()
+doubleSchubertPolynomial(List) := (w) -> (
+    n := #w;
+    x := local x;
+    y := local y;
+    Q := QQ[x_1..x_n,y_1..y_n];
+    schubertPolyHelper(w, Q, Double=>true)
+)
+
+
+
+dividedDifference = method(Options=>{Operator=>null})
+dividedDifference (RingElement, ZZ) := opts-> (f,i) -> (
+    Q:= ring f;
+    sf := sub(f, {Q_(i-1)=>Q_i, Q_i=>Q_(i-1)});
+    if opts.Operator === null then 
+        (f-sf) // (Q_(i-1)-Q_i)
+    else if opts.Operator == "Grothendieck" then
+        dividedDifference((1-Q_i)*f,i)
+    else error("Invalid option for operator.")
+    )
+
+
+polyByDividedDifference = method(Options=>{PolyType=>"Schubert"})
+polyByDividedDifference (List, PolynomialRing) := opts -> (w, Q) -> (
+    n := #w;
+    w0 := longestPerm(n);
+    schubw0 := product(n, i->(Q_i)^(n-1-i));
+    curpoly := schubw0;
+    if (w0 == w) then schubw0
+    else (
+        v := composePerms(inverseOf(w), w0);
+        redword := getOneReducedWord(v);
+	if opts.PolyType == "Schubert" then
+            polys := apply(reverse redword, i-> (curpoly=dividedDifference(curpoly,i); curpoly))
+	else if opts.PolyType == "Grothendieck" then
+	    polys = apply(reverse redword, i-> (curpoly=dividedDifference(curpoly,i, Operator=>"Grothendieck"); curpoly))
+	else error ("Invalid option for this polynomial type.");
+	polys#(#polys-1)
+    )
+    )
+
+
+------------------------------------------------
+------------------------------------------------
+--Pipe Dreams
+--(Implementation based on code of Allen Knutson)
+--Sample Usage:  pipeDreams({1,4,3,2})
+------------------------------------------------
+------------------------------------------------
+
+PipeDream = new SelfInitializingType of BasicList
+PipeDream.synonym = "pipe dream"
+
+-- for testing
+PipeDream == List := (D,L) -> toList D == L
+
+-- for printing
+expression       PipeDream := D -> netList(toList D, Boxes => false)
```

Comment:
the output of `expressoin` is supposed to be... an `Expression`. in the line above you should probably replace `experssion` with `net`?

Reply:
Yes, I intended this because there isn't a proper type of Expression for this class yet. Do you just want this instead?
```m2
expression       PipeDream := D -> expression netList(toList D, Boxes => false)
```